### PR TITLE
allow healing to prefer local disks over remote

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
allow healing to prefer local disks over remote

## Motivation and Context
optimization of healing, refer #17768 

## How to test this PR?
You need a setup with 4 servers and 4 disks in a 16-stripe setup.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/erasure-decode.go [hunk 1]
```diff
@@ -289,12 +289,15 @@ func (e Erasure) Decode(ctx context.Context, writer io.Writer, readers []io.Read
...
 
 // Heal reads from readers, reconstruct shards and writes the data to the writers.
-func (e Erasure) Heal(ctx context.Context, writers []io.Writer, readers []io.ReaderAt, totalLength int64) (derr error) {
+func (e Erasure) Heal(ctx context.Context, writers []io.Writer, readers []io.ReaderAt, totalLength int64, prefer []bool) (derr error) {
 	if len(writers) != e.parityBlocks+e.dataBlocks {
 		return errInvalidArgument
 	}
 
 	reader := newParallelReader(readers, e, 0, totalLength)
+	if len(readers) == len(prefer) {
```
Comment (jiuker) [inline cmd/erasure-decode.go:298]: There is always true. That's on purpose?
```go
readers := make([]io.ReaderAt, len(latestDisks))
prefer := make([]bool, len(latestDisks))
```

**Review by jiuker — CHANGES_REQUESTED** [review]

File: cmd/erasure-decode.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/erasure-decode.go:298]: yes it checks if the input `prefer` is non-nil and it matches number of readers.  @jiuker

Comment (jiuker) [reply cmd/erasure-decode.go:298]: @harshavardhana I mean We should check the `prefer`'s `true` length?

Comment (harshavardhana) [reply cmd/erasure-decode.go:298]: no not really @jiuker it must be same as available readers if its not then we do not enable preferential reading.

**Review by jiuker — APPROVED** [review]

**Review by klauspost — APPROVED** [review]
LGTM

**Pull request merged**